### ISSUE TITLE
do not show props injected by the framework in TS interfaces for field (and other) components

### ISSUE THREAD

**Issue by bayareacoder** [marmelab/react-admin#5500]
Field components receive injected props from their parent, eg. in
```
<Datagrid>
   <TextField label="Address" source="address" />
</Datagrid>
```
The TextField receives: `basePath`, `record` and `resource` as it is being cloned by the Datagrid iterator.
To avoid a TS error for missing props in the above, the TextField props are defined to have the injected props as optional:
```
declare const TextField: FC<TextFieldProps>;
export interface TextFieldProps extends PublicFieldProps, InjectedFieldProps, TypographyProps {
    component?: ElementType<any>;
}

export interface InjectedFieldProps<RecordType extends Record = Record> {
    basePath?: string;
    record?: RecordType;
    resource?: string;
}
```
However this is not really correct: injected props are not optional, they are required but guaranteed to be provided by the  framework (as long as the component is correctly used). 

So I believe a better typing would be to only require the public props (no need to mention the injected props in user code) but have the injected props non optional:
```
declare const TextField: FC<TextFieldProps>;
export interface TextFieldProps extends PublicFieldProps, TypographyProps {
    component?: ElementType<any>;
}
export interface InjectedFieldProps<RecordType extends Record = Record> {
    basePath: string;
    record: RecordType;
    resource: string;
}
```
Then still access injected props inside the component as follows:
```
const { basePath, record, resource} = props as InjectedFieldProps;
```

This restricts the interface to only showing public props.

Then custom field components would just require the same one-liner above to make it all work without cluttering their interface. Another benefit is that VSCode will then only show the public props. Currently it is not clear from code completion if a prop is injected automatically or is optional in user code. And if a user supplies an injected prop, it will actually not be taken into account (as the injected prop from cloning will take priority) - could be a hard to find error.  With this change it will not be possible to provide a prop that is injected from user code.

PS
I guess there would still an issue when using the component eg inside a `FormDataConsumer`where props are not injected automatically...Hmm...

**Comment by fzaninotto** [comment]
Hi @bayareacoder, and thanks for your suggestion. We're aware of the problems our TypeScript types cause due to props injection by the parent. But we've decided to work on a different fix than what you're suggesting.

In the next major version, we won't clone children and inject props anymore. Instead, to get information about the current record, list, or resource, each component will be able to call a context hook. So we'll switch from a "push" strategy to a "pull" strategy.

We've already started implementing this strategy in the main controllers (see useListContext, useEditContext, etc). We're doing it in a backwards-compatible way, so we keep on injecting props for now. As a consequence, we must keep the old types, too.

So this problem will be fixed when we release the next major. It won't come overnight - we're depending on material-ui to release their major version, too. In the meantime, you'll have to use the current types, with their imperfections. But that's better than no types, don't you think?
